saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/serviceprovider/registration/RelyingPartyRegistration.java

```diff
@@ -0,0 +1,205 @@
+ *		//remote WebSSO Endpoint - Where to Send AuthNRequests to
+ *		String webSsoEndpoint = "https://simplesaml-for-spring-saml.cfapps.io/saml2/idp/SSOService.php";
+ *		//local registration ID
+ *		String registrationId = "simplesamlphp";
+ *		//local entity ID - autogenerated based on URL
+ *		String localEntityIdTemplate = "{baseUrl}/saml2/service-provider-metadata/{registrationId}";
+ *		//local signing (and local decryption key and remote encryption certificate)
+ *		Saml2X509Credential signingCredential = getSigningCredential();
+ *		//IDP certificate for verification of incoming messages
+ *		Saml2X509Credential idpVerificationCertificate = getVerificationCertificate();
+ *		RelyingPartyRegistration rp = RelyingPartyRegistration.withRegistrationId(registrationId)
+ * 				.remoteIdpEntityId(idpEntityId)
+ * 				.idpWebSsoUrl(webSsoEndpoint)
+ * 				.credential(signingCredential)
+ * 				.credential(idpVerificationCertificate)
+ * 				.localEntityIdTemplate(localEntityIdTemplate)
+ * 				.build();
+ * </pre>
+ * @since 5.2
+ */
+public class RelyingPartyRegistration {
+
+	private final String registrationId;
+	private final String remoteIdpEntityId;
+	private final String assertionConsumerServiceUrlTemplate;
+	private final String idpWebSsoUrl;
+	private final List<Saml2X509Credential> credentials;
+	private final String localEntityIdTemplate;
+
+	private RelyingPartyRegistration(String idpEntityId, String registrationId, String assertionConsumerServiceUrlTemplate,
+			String idpWebSsoUri, List<Saml2X509Credential> credentials, String localEntityIdTemplate) {
+		hasText(idpEntityId, "idpEntityId cannot be empty");
+		hasText(registrationId, "registrationId cannot be empty");
+		hasText(assertionConsumerServiceUrlTemplate, "assertionConsumerServiceUrlTemplate cannot be empty");
+		hasText(localEntityIdTemplate, "localEntityIdTemplate cannot be empty");
+		notEmpty(credentials, "credentials cannot be empty");
+		notNull(idpWebSsoUri, "idpWebSsoUri cannot be empty");
+		for (Saml2X509Credential c : credentials) {
+			notNull(c, "credentials cannot contain null elements");
+		}
+		this.registrationId = registrationId;
+		this.remoteIdpEntityId = idpEntityId;
+		this.assertionConsumerServiceUrlTemplate = assertionConsumerServiceUrlTemplate;
+		this.credentials = new LinkedList<>(credentials);
+		this.idpWebSsoUrl = idpWebSsoUri;
+		this.localEntityIdTemplate = localEntityIdTemplate;
+	}
+
+	public String getRemoteIdpEntityId() {
+		return this.remoteIdpEntityId;
+	}
+
+	public String getRegistrationId() {
+		return this.registrationId;
+	}
+
+	public String getAssertionConsumerServiceUrlTemplate() {
+		return this.assertionConsumerServiceUrlTemplate;
+	}
+
+	public String getIdpWebSsoUrl() {
+		return this.idpWebSsoUrl;
+	}
+
+	public String getLocalEntityIdTemplate() {
+		return this.localEntityIdTemplate;
+	}
+
+	public List<Saml2X509Credential> getCredentialsForUsage(Saml2X509CredentialUsage... types) {
```

Comment:
It's a little weird to have to call `getCredentialsForUsage()` to just get the list of creds. It also seems pretty uncommon that I'd want to filter by multiple types.

Let's please add `getCredentials()` and use for that purpose.

And let's please simplify this so that it doesn't take a varargs and rename to `getCredentials()`.

Further, I think it would align better to do `getSigningCredentials()`, `getVerifyingCredentials()`, etc. since we have `isSigningCredential()`, `isVerifyingCredential()`, etc. on `Saml2X509Credential`.

Reply:
Better names used
00505f36